### PR TITLE
Story/cite 91

### PR DESCRIPTION
**Pull request by MaulikLim**
# Guidelines for Pull Requests

If you haven't yet read our code review guidelines, please do so, You can find them [here](https://diging.atlassian.net/wiki/spaces/DIGING/pages/2256076801/Code+Review+Guidelines).

Please confirm the following by adding an x for each item (turn `[ ]` into `[x]`).

- [x] I have removed all code style changes that are not necessary (e.g. changing blanks across the whole file that don’t need to be changed, adding empty lines in parts other than your own code)
- [x] I am not making any changes to files that don’t have any effect (e.g. imports added that don’t need to be added)
- [x] I do not have any sysout statements in my code or commented out code that isn’t needed anymore
- [x] I am not reformatting any files in the wrong format or without cause. 
- [x] I am not changing file encoding or line endings to something else than UTF-8, LF
- [x] My pull request does not show an insane amount of files being changed although my ticket only requires a few files being changed
- [x] I have added Javadoc/documentation where appropriate
- [x] I have added test cases where appropriate
- [x] I have explained any part of my code/implementation decisions that is not be self-explanatory

## Please provide a brief description of your ticket 
(you can copy the ticket if it hasn't changed) 

[CITE-91] - It should be possible to revert a citation to a previous version
[CITE-90] - It should be possible to see previous versions of a citation
  
## Anything else the reviewer needs to know?

Combined pull request for 2 tickets

[CITE-91]: https://diging.atlassian.net/browse/CITE-91?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: citesphere/src/main/java/edu/asu/diging/citesphere/web/user/RevertCitationController.java [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+            @PathVariable("zoteroGroupId") String zoteroGroupId, @PathVariable("itemId") String itemId,
+            @PathVariable("version") Long version) {
+
+        try {
+            citationVersionManager.revertCitationVersion((IUser) authentication.getPrincipal(), zoteroGroupId, itemId,
+                    version);
+            return "redirect:/auth/group/" + zoteroGroupId + "/items/" + itemId;
+        } catch (GroupDoesNotExistException | ZoteroConnectionException | CitationIsOutdatedException
+                | ZoteroHttpStatusException | CannotFindCitationVersionException | CannotFindCitationException e) {
+            logger.error(e.getMessage());
```
Comment (jdamerow) [inline citesphere/src/main/java/edu/asu/diging/citesphere/web/user/RevertCitationController.java:40]: log complete stack trace

File: citesphere/src/main/java/edu/asu/diging/citesphere/web/user/RevertCitationController.java [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+import edu.asu.diging.citesphere.core.service.ICitationVersionManager;
+import edu.asu.diging.citesphere.user.IUser;
+
+@Controller
+public class RevertCitationController {
+    
+    private final Logger logger = LoggerFactory.getLogger(getClass());
+
+    @Autowired
+    ICitationVersionManager citationVersionManager;
```
Comment (jdamerow) [inline citesphere/src/main/java/edu/asu/diging/citesphere/web/user/RevertCitationController.java:27]: should be private

File: citesphere/src/main/java/edu/asu/diging/citesphere/core/service/impl/CitationVersionManager.java [hunk 3]
```diff
@@ -0,0 +1,115 @@
...
+     * java.lang.String, java.lang.String, int, int)
+     */
+    @Override
+    public List<CitationVersion> getCitationVersions(IUser user, String groupId, String key, int page, int pageSize)
+            throws GroupDoesNotExistException {
+        ICitationGroup group;
+        try {
+            group = groupManager.getGroup(user, groupId);
+        } catch (HttpClientErrorException ex) {
+            throw new GroupDoesNotExistException("Group with id " + groupId + " does not exist.");
```
[the remark this answers is not in the document]
Comment (jdamerow) [reply citesphere/src/main/java/edu/asu/diging/citesphere/core/service/impl/CitationVersionManager.java:54]: `ex` should be passed into the new exception to keep a valid stack trace

File: citesphere/src/main/java/edu/asu/diging/citesphere/core/service/impl/CitationVersionManager.java [hunk 4]
```diff
@@ -0,0 +1,115 @@
...
+     * java.lang.String, java.lang.String, java.lang.Long)
+     */
+    @Override
+    public ICitation getCitationVersion(IUser user, String groupId, String key, Long version)
+            throws GroupDoesNotExistException {
+        ICitationGroup group;
+        try {
+            group = groupManager.getGroup(user, groupId);
+        } catch (HttpClientErrorException ex) {
+            throw new GroupDoesNotExistException("Group with id " + groupId + " does not exist.");
```
[the remark this answers is not in the document]
Comment (jdamerow) [reply citesphere/src/main/java/edu/asu/diging/citesphere/core/service/impl/CitationVersionManager.java:85]: see above

**Review by jdamerow — CHANGES_REQUESTED** [review]

**Comment by jdamerow** [comment]
Make it so, Jenkins.

**Comment by jdamerow** [comment]
The original version never seems to be stored. The version that is first stored is the one after it has been changed. Also, the version page that lists all versions should also include the time of a version (not just the date)

**Comment by jdamerow** [comment]
Make it so, Jenkins.

**Comment by jdamerow** [comment]
- the updated by column always contains "unauthenticated"
- the files portion of the items page (for a previous version) is empty

**Comment by MaulikLim** [comment]
image


> * the updated by column always contains "unauthenticated"
> * the files portion of the items page (for a previous version) is empty

- It shows "unauthenticated" when the item is updated through sync. The author of the generated history is provided by Spring security. For me, it shows my username when I edit the item.
- Fixed the files portion.

**Comment by jdamerow** [comment]
Then let's make it "Sync Process" instead of "unauthenticated"

**Comment by MaulikLim** [comment]
> Then let's make it "Sync Process" instead of "unauthenticated"

Done

**Review by jdamerow — CHANGES_REQUESTED** [review]

**Comment by jdamerow** [comment]
Make it so, Jenkins.

**Comment by jdamerow** [comment]
Make it so, Jenkins.

**Pull request merged**